pkg/ha/stage.go

```diff
@@ -0,0 +1,364 @@
+	defer cancel()
+
+	var stage Stage
+	resp, err := cli.Get(ctx, common.StageRelayKeyAdapter.Encode(source))
+	if err != nil {
+		return stage, 0, err
+	}
+
+	if resp.Count == 0 {
+		return stage, 0, nil
+	} else if resp.Count > 1 {
+		// TODO(csuzhangxc): add terror.
+		// this should not happen.
+		return stage, 0, fmt.Errorf("too many relay stage (%d) exist for source %s", resp.Count, source)
+	}
+
+	stage, err = stageFromJSON(string(resp.Kvs[0].Value))
+	if err != nil {
+		return stage, 0, err
+	}
+
+	return stage, resp.Header.Revision, nil
+}
+
+// GetAllRelayStage gets all relay stages.
+// k/v: source ID -> relay stage.
+func GetAllRelayStage(cli *clientv3.Client) (map[string]Stage, int64, error) {
+	ctx, cancel := context.WithTimeout(cli.Ctx(), etcdutil.DefaultRequestTimeout)
+	defer cancel()
+
+	resp, err := cli.Get(ctx, common.StageRelayKeyAdapter.Path(), clientv3.WithPrefix())
+	if err != nil {
+		return nil, 0, err
+	}
+
+	stages := make(map[string]Stage)
+	for _, kv := range resp.Kvs {
+		stage, err2 := stageFromJSON(string(kv.Value))
+		if err2 != nil {
+			return nil, 0, err2
+		}
+		stages[stage.Source] = stage
+	}
+	return stages, resp.Header.Revision, nil
+}
+
+// GetSubTaskStage gets the subtask stage for the specified upstream source and task name.
+// if the stage for the source and task name not exist, return with `err == nil` and `revision=0`.
+// if task name is "", it will return all subtasks' stage as a map{task-name: stage} for the source.
+// if task name is given, it will return a map{task-name: stage} whose length is 1.
+func GetSubTaskStage(cli *clientv3.Client, source, task string) (map[string]Stage, int64, error) {
+	ctx, cancel := context.WithTimeout(cli.Ctx(), etcdutil.DefaultRequestTimeout)
+	defer cancel()
+
+	var (
+		stm  = make(map[string]Stage)
+		resp *clientv3.GetResponse
+		err  error
+	)
+	if task != "" {
+		resp, err = cli.Get(ctx, common.StageSubTaskKeyAdapter.Encode(source, task))
+	} else {
+		resp, err = cli.Get(ctx, common.StageSubTaskKeyAdapter.Encode(source), clientv3.WithPrefix())
+	}
+
+	if err != nil {
+		return stm, 0, err
+	}
+
+	if resp.Count == 0 {
+		return stm, 0, nil
+	} else if task != "" && resp.Count > 1 {
+		return stm, 0, fmt.Errorf("too many stage (%d) exist for subtask {sourceID: %s, task name: %s}", resp.Count, source, task)
+	}
+
+	for _, kvs := range resp.Kvs {
+		stage, err2 := stageFromJSON(string(kvs.Value))
+		if err2 != nil {
+			return stm, 0, err2
+		}
+		stm[stage.Task] = stage
+	}
+
+	return stm, resp.Header.Revision, nil
+}
+
+// GetAllSubTaskStage gets all subtask stages.
+// k/v: source ID -> task name -> subtask stage.
+func GetAllSubTaskStage(cli *clientv3.Client) (map[string]map[string]Stage, int64, error) {
+	ctx, cancel := context.WithTimeout(cli.Ctx(), etcdutil.DefaultRequestTimeout)
+	defer cancel()
+
+	resp, err := cli.Get(ctx, common.StageSubTaskKeyAdapter.Path(), clientv3.WithPrefix())
+
```

Comment:
can delete this line

Reply:
deleted in https://github.com/pingcap/dm/pull/473/commits/73cee97e06a098dd821d97b742e3b0556bac8f19.